src/domain/shared/Log.ts

```diff
@@ -0,0 +1,245 @@
+        const key = `${messageType || "undefined"} | ${level} | ${message}`;
+
+        const id = Logger.#_getMessageId(idMap, key);
+
+        if (!this.#_messageFlags.get(id)) {
+            this.#_messageFlags.set(id, true);
+            this.level(level, message, messageType);
+        }
+    }
+
+    /*@devdoc
+     *  Logs a message at the default level, with optional type.
+     *  @param {message} string - the message to log
+     *  @param {messageType} string - optional user defined message type
+     */
+    message(message: string, messageType?: string): void {
+        this.level(LogLevel.DEFAULT, message, messageType);
+    }
+
+    /*@devdoc
+     *  Logs a message at the info level, with optional type.
+     *  @param {message} string - the message to log
+     *  @param {messageType} string - optional user defined message type
+     */
+    info(message: string, messageType?: string): void {
+        this.level(LogLevel.INFO, message, messageType);
+    }
+
+    /*@devdoc
+     *  Logs a message at the debug level, with optional type.
+     *  @param {message} string - the message to log
+     *  @param {messageType} string - optional user defined message type
+     */
+    debug(message: string, messageType?: string): void {
+        this.level(LogLevel.DEBUG, message, messageType);
+    }
+
+    /*@devdoc
+     *  Logs a message at the warning level, with optional type.
+     *  @param {message} string - the message to log
+     *  @param {messageType} string - optional user defined message type
+     */
+    warning(message: string, messageType?: string): void {
+        this.level(LogLevel.WARNING, message, messageType);
+    }
+
+    /*@devdoc
+     *  Logs a message at the error level, with optional type.
+     *  @param {message} string - the message to log
+     *  @param {messageType} string - optional user defined message type
+     */
+    error(message: string, messageType?: string): void {
+        this.level(LogLevel.ERROR, message, messageType);
+    }
+
+    /*@devdoc
+     *  Applies a filer to all subsequently logged messages, based on level.
+     *  @param {pred} (level: LogLevel) => boolean - the filtering condition
+     */
+    filterLevels(pred: (level: LogLevel) => boolean): void {
+        this.#_activeFunctions = new Map<LogLevel, LogFunction>();
+        for (const level of allLogLevels) {
+            const func = this.#_context.getFunction(level);
+            if (func && pred(level)) {
+                this.#_activeFunctions.set(level, func);
+            }
+        }
+    }
+
+    /*@devdoc
+     *  Sets a filter for all subsequently logged messages, based on user defined types.
+     *  @param {filter} Array<string | undefined> - the collection of types to allow,
+     *  presence or presence of undefined value will determine whether to show or hide messages with no type specified.
+     */
+    setTypeFilter(filter?: Array<string | undefined>): void {
+        this.#_typeFilter = filter;
+    }
```

Comment:
I wonder if the filter could just be a string that is searched for in logged messages rather than requiring a special `messageType` parameter. Doing this would enable logging of more than a single string.

One step further could be filtering using a string | RegExp.

Reply:
I used an optional parameter for message to keep things simple initially, but I see how it can be annoying for various reason. I think a better solution for it would be to use method chaining:
```
Log.type("audio").warning(...);
```
Then you could also define the typed logger once (maybe in some top level module where it makes sense, or even in Log.ts for commonly used types) and not have to specify the type as a string everywhere. 

We can additionally have string and regexp filtering, but that won't be a good replacement for explicit message types I think. We can kind of substitute it with a convention (like specify type in square bracket at the very beginning) and use regex to filter, but then we have to worry about strictly following the convention at the call site, and if we ever want to change the convention we would have to modify all the calls. With an explicit type the conventions of how the log level and type are presented is specified just once in logger context, which also allows for some flexibility, for example the string logger context puts the log level after the type.